File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    answer = 100
    for i in range(n-1):
        wire = wires.copy()
        tmp = wire.pop(i)
        q1,q2 = deque([tmp[0]]),deque([tmp[1]])
        ch1,ch2 = [tmp[0]],[tmp[1]]
        dic = {x:[] for x in range(1,n+1)}
        for x in wire:
            dic[x[0]].append(x[1])
            dic[x[1]].append(x[0])
        while(q1):
            cm1 = q1.popleft()
            for j in dic[cm1]:
                if j not in ch1:
                    q1.append(j)
                    ch1.append(j)
        while(q2):
            cm2 = q2.popleft()
            for j in dic[cm2]:
                if j not in ch2:
                    q2.append(j)
                    ch2.append(j)
        ln1,ln2 = len(ch1),len(ch2)
        print(ch1,ch2)
        answer = min(abs(ln2-ln1),answer)
    return answer

File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        for (int i = 0; i<wires.length; i++){  
            HashMap<Integer,ArrayList<Integer>> graph = new HashMap();
            for (int j = 0; j < n+1; j++) graph.put(j+1,new ArrayList<Integer>());
            
            for (int j = 0; j < wires.length; j++){
                if (i ==j) continue;
                int[] wire = wires[j];
                graph.get(wire[0]).add(wire[1]);
                graph.get(wire[1]).add(wire[0]);
            }
            
            boolean[] visit = new boolean[n+1];
            
            int node_a = wires[i][0];
            visit[node_a] = true;
            Queue<Integer> queue_a = new LinkedList();
            queue_a.add(node_a);
            int cnt_a = 1;
            while(!queue_a.isEmpty()){
                int node = queue_a.poll();
                for (int nextNode: graph.get(node)){
                    if (!visit[nextNode]){
                        queue_a.add(nextNode);
                        cnt_a++;
                        visit[nextNode] = true;
                    }
                }
            }
            
            int node_b = wires[i][1];
            visit[node_b] = true;
            Queue<Integer> queue_b = new LinkedList();
            queue_b.add(node_b);
            int cnt_b = 1;
            while(!queue_b.isEmpty()){
                int node = queue_b.poll();
                for (int nextNode: graph.get(node)){
                    if (!visit[nextNode]){
                        queue_b.add(nextNode);
                        cnt_b++;
                        visit[nextNode] = true;
                    }
                }
            }
            
            answer = Math.min(answer,Math.abs(cnt_a-cnt_b));
        }
        return answer;
    }
}